### PR TITLE
Check for hardcoded color and fonts on nib files

### PR DESCRIPTION
**Pull request by eiditakahashi**
- Check for hardcoded fonts and colors on nib files

### CODE REVIEW

**Comment by eiditakahashi** [comment]
Tested on a PR and result was:
`<color key="backgroundColor" red="1" green="0.99997437000274658" blue="0.99999129772186279" alpha="1" colorSpace="custom" customColorSpace="sRGB"/>`
`<color key="backgroundColor" red="1" green="0.99997437000274658" blue="0.99999129772186279" alpha="1" colorSpace="custom" customColorSpace="sRGB"/>`
`<color key="backgroundColor" red="0.0" green="0.0" blue="0.0" alpha="0.0" colorSpace="custom" customColorSpace="sRGB"/>`

@ghashi can you confirm if regex is getting the correct result?

**Comment by ghashi** [comment]
@eiditakahashi Just the last color seems to be "black", therefore shouldn't be there. The others seems to be "hardcoded".

ps: this is the script on "Partograma"
``` bash
function find_hardcoded_colors {
  color_count=$(find $project_path -name *.xib -not -path "$project_path/.git/*" -not -path "$project_path/Pods/*" -print0 | xargs -0 grep -i 'color' \
    | grep -v 'white="1"' \
    | grep -v 'white="0.0"' \
    | grep -v 'red="0.93725490196078431" green="0.93725490196078431" blue="0.95686274509803926" alpha="1"' \
    | grep -v 'red="0.93725490199999995" green="0.93725490199999995" blue="0.95686274510000002" alpha="1"' \
    | grep -v 'red="0.0" green="0.0" blue="0.0"' \
    | grep -v 'red="1" green="1" blue="1" alpha="1"' \
    | grep -v '<color key="textColor" cocoaTouchSystemColor="darkTextColor"/>' \
    | grep -v '<color key="titleShadowColor"' \
    | grep -v 'cocoaTouchSystemColor' \
    | grep -v '<nil' \
    | grep -v '<view' \
    | grep -v 'userDefinedRuntimeAttribute' \
    | grep -v '<string key="text"' \
    | grep -v 'image' \
    | grep -v '</string>' \
    | grep -v '<label' \
    | grep -v '<document' \
    | wc -l)
  assert_message "Number of hardcoded COLORS:" $expected_hardcoded_colors $color_count
}
```

ps2: this part of the script `grep -v 'red="0.0" green="0.0" blue="0.0"'` should remove the last color `<color key="backgroundColor" red="0.0" green="0.0" blue="0.0" alpha="0.0" colorSpace="custom" customColorSpace="sRGB"/>` , shouldn't?

**Comment by eiditakahashi** [comment]
@ghashi When I created the regex I was using an old version of this script and this line was different. Updating the line removed the last color! Thanks!

**Review by felipesabino — APPROVED** [review]

File: Dangerfile [hunk 1]
```diff
@@ -296,6 +297,22 @@ def validateSwiftFiles(file)
   end
 end
 
+def checkHardcodedNib(file)
+  File.foreach(file) do |line|
+    line = line.gsub('\n','').strip
+    ext = File.extname(file)
+    case ext
+    when ".xib"
+      if line =~ /<color(?!.*(?:white="1"|white="0\.0"|red="0\.93725490196078431" green="0\.93725490196078431" blue="0\.95686274509803926" alpha="1"|red="0\.93725490199999995" green="0\.93725490199999995" blue="0\.95686274510000002" alpha="1"|red="0\.0" green="0\.0" blue="0\.0"|red="1" green="1" blue="1" alpha="1"|key="textColor" cocoaTouchSystemColor="darkTextColor"\/>|key="titleShadowColor"|cocoaTouchSystemColor|<nil|<view|userDefinedRuntimeAttribute|<string key="text"|image|<\/string>|<label|<document)).*\/>/
```
Comment (felipesabino) [inline Dangerfile]: @ghashi just out of curiosity, how did you come up with these values like `0. 95686274509803926`? Are they part of XCode crazy behavior to somehow add them when hardcoded values are present or were these vales specific for some project?

Comment (ghashi) [reply Dangerfile]: @felipesabino 
Don't you know `red="0.93725490196078431" green="0.93725490196078431" blue="0.95686274509803926"` ? 
It's my favourite color! hahahah

yeap... These values are some default colors set by XCode. Depending on the view you add, they'll have a gray color which has this hex =/

Comment (melloflavio) [reply Dangerfile]: @ghashi @eiditakahashi 
Any way we can break this into smaller declarations? This regex as it is now would be nearly impossible to debug and maintain.

I was thinking something along the lines of:
```ruby
COLOR_WHITE_1                  = white="1"
COLOR_WHITE_0                  = white="0\.0"
COLOR_GRAY_ARBITRARY_1         = red="0\.93725490196078431" green="0\.93725490196078431" blue="0\.95686274509803926" alpha="1"
COLOR_GRAY_ARBITRARY_2         = red="0\.93725490199999995" green="0\.93725490199999995" blue="0\.95686274510000002" alpha="1"
COLOR_BLACK                    = red="0\.0" green="0\.0" blue="0\.0"
COLOR_WHITE_2                  = red="1" green="1" blue="1" alpha="1"
ATTR_TEXT_COLOR                = key="textColor" cocoaTouchSystemColor="darkTextColor"\/>
ATTR_TITLE_SHADOW_COLOR        = key="titleShadowColor"
COCOA_TOUCH_SYSTEM_COLOR       = cocoaTouchSystemColor
TAG_NIL                        = <nil
TAG VIEW                       = <view
USER_DEFINED_RUNTIME_ATTRIBUTE = userDefinedRuntimeAttribute
TAG_STRING                     = <string key="text"
IMAGE                          = image
TAG_STRING_EMPTY               = <\/string>
TAG_LABEL                      = <label
TAG_DOCUMENT                   = <document


if line =~ /
<color(
  ?!.*(
    ?:#{Regexp.quote(COLOR_WHITE_1)}
    | #{Regexp.quote(COLOR_WHITE_0)}
    | #{Regexp.quote(COLOR_GRAY_ARBITRARY_1)}
    | #{Regexp.quote(COLOR_GRAY_ARBITRARY_2)}
    | #{Regexp.quote(COLOR_BLACK)}
    | #{Regexp.quote(COLOR_WHITE_2)}
    | #{Regexp.quote(ATTR_TEXT_COLOR)}
    | #{Regexp.quote(ATTR_TITLE_SHADOW_COLOR)}
    | #{Regexp.quote(COCOA_TOUCH_SYSTEM_COLOR)}
    | #{Regexp.quote(TAG_NIL)}
    | #{Regexp.quote(TAG)}
    | #{Regexp.quote(USER_DEFINED_RUNTIME_ATTRIBUTE)}
    | #{Regexp.quote(TAG_STRING)}
    | #{Regexp.quote(IMAGE)}
    | #{Regexp.quote(TAG_STRING_EMPTY)}
    | #{Regexp.quote(TAG_LABEL)}
    | #{Regexp.quote(TAG_DOCUMENT)}
  )
)
.*\/>/
```
At the very least, [breaking the regex into multiple lines](http://stackoverflow.com/questions/3762183/split-ruby-regex-over-multiple-lines) would really help the next person to understand what's being done.

```ruby
if line =~ /
<color(
  ?!.*(
    ?:white="1"                                                                                     #COLOR_WHITE_1
    | white="0\.0"                                                                                  #COLOR_WHITE_0
    | red="0\.93725490196078431" green="0\.93725490196078431" blue="0\.95686274509803926" alpha="1" #COLOR_GRAY_ARBITRARY_1
    | red="0\.93725490199999995" green="0\.93725490199999995" blue="0\.95686274510000002" alpha="1" #COLOR_GRAY_ARBITRARY_2
    | red="0\.0" green="0\.0" blue="0\.0"                                                           #COLOR_BLACK
    | red="1" green="1" blue="1" alpha="1"                                                          #COLOR_WHITE_2
    | key="textColor" cocoaTouchSystemColor="darkTextColor"\/>                                      #ATTR_TEXT_COLOR
    | key="titleShadowColor"                                                                        #ATTR_TITLE_SHADOW_COLOR
    | cocoaTouchSystemColor                                                                         #COCOA_TOUCH_SYSTEM_COLOR
    | <nil                                                                                          #TAG_NIL
    | <view                                                                                         #TAG
    | userDefinedRuntimeAttribute                                                                   #USER_DEFINED_RUNTIME_ATTRIBUTE
    | <string key="text"                                                                            #TAG_STRING
    | image                                                                                         #IMAGE
    | <\/string>                                                                                    #TAG_STRING_EMPTY
    | <label                                                                                        #TAG_LABEL
    | <document                                                                                     #TAG_DOCUMENT
  )
)
.*\/>/
```

Comment (eiditakahashi) [reply Dangerfile]: @melloflavio agreed. Refactored and tested! 👍

**Review by melloflavio — APPROVED** [review]
👍

**Pull request merged**